src/main/java/com/minecolonies/coremod/colony/ColonyList.java

```diff
@@ -0,0 +1,183 @@
+    {
+        T existingColony = get(colony.getID());
+        if (existingColony != null && existingColony != colony) {
+            throw new IllegalArgumentException(
+                    String.format("Already a colony registered to id=%d, colony=%s, not changing to colony=%s",
+                            colony.getID(),
+                            existingColony.getName(),
+                            colony.getName()));
+        }
+
+        list[colony.getID()] = colony;
+    }
+
+    public void remove(Colony colony)
+    {
+        remove(colony.getID());
+    }
+
+    public void remove(int id)
+    {
+        if (list[id] == null) {
+            Log.getLogger().warn("Tried to remove colony with id=%d, but it didn't exist.", id);
+        }
+
+        size--;
+        list[id] = null;
+
+        if (!nullIndices.contains(id)) {
+            nullIndices.add(id);
+        }
+    }
+
+    private int getNextColonyID()
+    {
+        if (nullIndices.isEmpty()) {
+            topID++;
+            if (topID >= list.length) {
+                // Expand list
+                Object[] newList = new Object[list.length * 2];
+                System.arraycopy(list, 0, newList, 0, list.length);
+                list = newList;
+            }
+
+            return topID;
+        }
+
+        return nullIndices.remove(0);
+    }
+
+    public void clear()
+    {
+        for (int i = 0; i < list.length; i++)
+        {
+            list[i] = null;
+        }
+
+        nullIndices.clear();
+
+        topID = 0;
+        size = 0;
+    }
+
+    public int size()
+    {
+        return size;
+    }
+
+    public boolean isEmpty()
+    {
+        return size == 0;
+    }
+
+    @NotNull
+    public List<T> getCopyAsList()
+    {
+        final List<T> copyList = new ArrayList<>();
+        for (final T colony : this) {
+            copyList.add(colony);
+        }
+
+        return copyList;
+    }
+
+    @Override
+    public Iterator<T> iterator()
+    {
+        return new Iterator<T>()
+        {
+
+            private int nextIndex = getNextIndex(0);
+
+            @Override
+            public boolean hasNext()
+            {
+                return nextIndex < list.length;
+            }
+
+            @Override
+            public T next()
+            {
+                final int index = nextIndex;
+                nextIndex = getNextIndex(nextIndex);
+                if (index >= list.length) {
+                    throw new NoSuchElementException();
+                }
+                return get(index);
+            }
+
+            private int getNextIndex(int startingIndex) {
+                int index = startingIndex + 1;
+                while (index < list.length) {
+                    if (list[index] != null) {
+                        return index;
+                    }
+
+                    index++;
+                }
+
+                return index;
+            }
+        };
+    }
+
+    public Stream<T> stream()
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Document this public method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUndocumentedApi)